### DISCUSSION TITLE
how to parse data from db to my dynamic page

### DISCUSSION THREAD

**Discussion by peterparker2005** [prisma/prisma#19476] [Q&A] [answered]
hello! i use prisma + postgres, nextjs and typescript. so the question is how to create a page of a [category] where will be shown items which have realtion to category by id. i think i need to export async func from my categories.tsx which is shown below, but what should i do in this function - idk... my schema: 
```
model Item {
  id         Int      @id @default(autoincrement())
  slug       String
  title      String
  about      String
  category   Category @relation(fields: [categoryId], references: [id])
  categoryId Int
  img_path   String
}

model Category {
  id       Int    @id @default(autoincrement())
  slug     String
  title    String
  items    Item[]
  img_path String
}
```
my funcs >>

```
import prisma from './client';

export async function getCategories() {
  try {
    const categoriesFound = await prisma.category.findMany();
    return { categoriesFound };
  } catch (error) {
    return { error };
  }
}

export async function getItemsForCategories() {
  try {
    const itemsForCategory = await prisma.category.findMany({
      where: {
        items: {
          every: {
            id: 1,
          },
        },
      },
    });
    return { itemsForCategory };
  } catch (error) {
    return { error };
  }
}

export async function createCategory(category: any) {
  try {
    const categoryCreated = await prisma.category.create({ data: category });
    return { category: categoryCreated };
  } catch (error) {
    return { error };
  }
}

export async function getCategoryById(id: number) {
  try {
    const category = await prisma.category.findUnique({ where: { id } });
    return { category };
  } catch (error) {
    return { error };
  }
}

export async function deleteCategories() {
  try {
    const deleteCategories = await prisma.category.deleteMany();
  } catch (error) {
    return { error };
  }
}

export async function updateCategory(id: number, data: any) {
  try {
    const updatedCategory = await prisma.category.update({
      where: { id },
      data,
    });
    return { category: updatedCategory };
  } catch (error) {
    return { error };
  }
}
```
and my api reqs >>

```
import { NextApiRequest, NextApiResponse } from 'next';
import {
  createCategory,
  getCategories,
  getCategoryById,
  deleteCategories,
  updateCategory,
} from '@/prisma/categories';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method === 'GET') {
    try {
      const categories = await getCategories();
      res.status(200).json({ categories });
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  } else if (req.method === 'POST') {
    try {
      const data = req.body;
      const category = await createCategory(data);
      res.status(201).json({ category });
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  } else if (req.method === 'DELETE') {
    try {
      await deleteCategories();
      return res.status(203).json({ message: 'Categories deleted' });
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  } else if (req.method === 'PATCH') {
    try {
      const { id, data } = req.body;
      const updatedCategory = await updateCategory(id, data);
      res.status(200).json({ item: updatedCategory });
    } catch (error) {
      res.status(500).json({ error: 'Internal server error' });
    }
  } else {
    res.status(404).json({ error: 'Not found' });
  }
}

```

**Comment by nurul3101** [comment, marked as answer]
Hello @coutp 👋 

> how to create a page of a [category] where will be shown items which have realtion to category by id

To create a page for a category, you need to fetch the category by its ID and display all items that belong to this category. I see that you already have a getCategoryById function that fetches a category by its ID. However, this function does not fetch the items associated with the category.

You can change your `getCategoryById` function to include the items as well like this:
```ts
export async function getCategoryById(id: number) {
  try {
    const category = await prisma.category.findUnique({
      where: { id },
      include: {
        items: true,
      },
    });
    return { category };
  } catch (error) {
    return { error };
  }
}
```

In this version of the function, `include: { items: true }` tells Prisma to fetch the items associated with the category.
You can use this function in your API handler. I'm assuming you want to fetch a category and its items when you send a GET request to `/api/categories/[id]` path.

**Reply by peterparker2005** [reply]
Hello @nurul3101, thank you so much for your help. I just wanted to clarify that I don't need to use an API; it was only for testing purposes. What I actually need is to fetch my items based on their category on the page `app/catalog/[slug]/page.tsx`. I'm a bit confused about how to use Prisma to render server components and get data for especially for dynamic pages. If you could assist me with this, I would greatly appreciate it.

Here's what I'm trying to accomplish:

Render fetched items based on their category.
Use the slug string from the category as the link for my `[slug]` page.

I'm not asking you to write the code for me, just give me advice or instruction on what I should read in the documentation or something else. should i use `useRouter()` or what... just cant get the logic how it works(

Please accept my apologies for taking up your time, and thank you sincerely for your assistance!
